### ISSUE TITLE
Faulty regex causes issues with pihole -q and consumes all memory until it's killed

### ISSUE THREAD

**Issue by yubiuser** [pi-hole/pi-hole#3173]
This has been [brought up at the v5.0 beta discourse forum](https://discourse.pi-hole.net/t/faulty-regex-consumes-all-memory-and-causes-issues-with-pihole-q/28710/20). Please see there for more details and screenshots.

Using a faulty regex like 

`(\.|^)*\.services\.generalmagic\.com`

causes issues when executing `pihole -q` resulting in dramatically memory increase until the init system kills the process. The output lists

`/opt/pihole/query.sh: line 37: 21500 Killed                  awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo "${lists}") <(echo "${domain}") 2> /dev/null`

Background: There is no validation of regex by pihole that it will ever match a domain. 

Expected behavior: I don’t think pihole should have a full-grown regex validator (In terms of ‘This is what the regex means’) but maybe a simple syntax checker ('This is not a valid regex und will never match any domain.) - especially as a non-valid regex can break pihole’s scripts. Or pihole should be more robust against faulty regex.

**Comment by pralor** [comment]
This issue has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/faulty-regex-consumes-all-memory-and-causes-issues-with-pihole-q/28710/30

**Comment by DL6ER** [comment]
> I don’t think pihole should have a full-grown regex validator (In terms of ‘This is what the regex means’) but maybe a simple syntax checker ('This is not a valid regex und will never match any domain.)

The second is not possible without the former. In your case, the issue arose from the sub-expression `^*` which translates into: An arbitrary amount of line endings. This **is** valid code, just somewhat meaningless.

> Or pihole should be more robust against faulty regex.

FTL is robust against such (meaningless) regex, it will short-circuit the impossible situation of an arbitrary amount of line beginnings per line to one. Hence, **normal Pi-hole operation should not be affected** by this regex. Did you experience any DNS issues?

However, it seems you have stumbled upon an original `awk` bug and I'm not sure whether we want/can address them, checking back with @dschaper

**Comment by yubiuser** [comment]
> Did you experience any DNS issues?

No DNS issues observed.

**Comment by dschaper** [comment]
I'm playing with a couple of ideas. Let me stew some more and I'll try to see if anything actually comes to a good option.

**Comment by dschaper** [comment]
Is `*` ever a valid quantifier for our purposes? Will there ever be a char or capture group that would be matched 0 or more times? Is it safe to say that anything to be matched should show up at least once?

**Comment by dschaper** [comment]
I don't know if this is really an `awk` bug. It seems to work as intended when used directly:

```sh
dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo '(\.|^)*\.services\.generalmagic\.com$') <(echo "gstatic.com")

dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo "(\.|^)*\.services\.generalmagic\.com$") <(echo "gstatic.com")

dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo "(\.|^)*\.services\.generalmagic\.com$") <(echo "generalmagic.com")

dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo "(\.|^)*\.services\.generalmagic\.com$") <(echo "services.generalmagic.com")

dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo "(\.|^)*\.services\.generalmagic\.com$") <(echo "me.services.generalmagic.com")
(\.|^)*\.services\.generalmagic\.com$

dschaper@Mariner-10:~$ awk 'NR==FNR{regexps[$0];next}{for (r in regexps)if($0 ~ r)print r}' <(echo '(\.|^)*\.services\.generalmagic\.com$') <(echo "me.services.generalmagic.com")
(\.|^)*\.services\.generalmagic\.com$
```

**Comment by yubiuser** [comment]
Seems to be an issue with certain versions of `awk`.
I can r[eplicate the issue with `mawk` but not with `gawk`](https://discourse.pi-hole.net/t/faulty-regex-consumes-all-memory-and-causes-issues-with-pihole-q/28710/46?u=yubiuser).

**Comment by DL6ER** [comment]
This issue is resolved via #3186
